// @vitejs/devtools-kit/dist/constants.js
//#region src/constants.ts
const DEVTOOLS_MOUNT_PATH = "/.devtools/";
const DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH = "/.devtools";
const DEVTOOLS_DIRNAME = ".devtools";
const DEVTOOLS_CONNECTION_META_FILENAME = ".connection.json";
const DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME = ".rpc-dump/index.json";
const DEVTOOLS_DOCK_IMPORTS_FILENAME = ".client-imports.js";
const DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID = "/.devtools-client-imports.js";
const DEVTOOLS_RPC_DUMP_DIRNAME = ".rpc-dump";
const DEFAULT_CATEGORIES_ORDER = {
	"~viteplus": -1e3,
	"default": 0,
	"app": 100,
	"framework": 200,
	"web": 300,
	"advanced": 400,
	"~builtin": 1e3
};
const DEFAULT_STATE_USER_SETTINGS = () => ({
	docksHidden: [],
	docksCategoriesHidden: [],
	docksPinned: [],
	docksCustomOrder: {},
	showIframeAddressBar: false,
	closeOnOutsideClick: false,
	commandShortcuts: {}
});
//#endregion
export { DEFAULT_CATEGORIES_ORDER, DEFAULT_STATE_USER_SETTINGS, DEVTOOLS_CONNECTION_META_FILENAME, DEVTOOLS_DIRNAME, DEVTOOLS_DOCK_IMPORTS_FILENAME, DEVTOOLS_DOCK_IMPORTS_VIRTUAL_ID, DEVTOOLS_MOUNT_PATH, DEVTOOLS_MOUNT_PATH_NO_TRAILING_SLASH, DEVTOOLS_RPC_DUMP_DIRNAME, DEVTOOLS_RPC_DUMP_MANIFEST_FILENAME };
