// @pnpm/logger/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeToConsole = exports.streamParser = exports.createStreamParser = exports.globalWarn = exports.globalInfo = exports.logger = void 0;
var logger_js_1 = require("./logger.js");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_js_1.logger; } });
Object.defineProperty(exports, "globalInfo", { enumerable: true, get: function () { return logger_js_1.globalInfo; } });
Object.defineProperty(exports, "globalWarn", { enumerable: true, get: function () { return logger_js_1.globalWarn; } });
var streamParser_js_1 = require("./streamParser.js");
Object.defineProperty(exports, "createStreamParser", { enumerable: true, get: function () { return streamParser_js_1.createStreamParser; } });
Object.defineProperty(exports, "streamParser", { enumerable: true, get: function () { return streamParser_js_1.streamParser; } });
var writeToConsole_js_1 = require("./writeToConsole.js");
Object.defineProperty(exports, "writeToConsole", { enumerable: true, get: function () { return writeToConsole_js_1.writeToConsole; } });
