// @vitejs/devtools/dist/context-internal-DZpoJjMz.js
import { debounce } from "perfect-debounce";
import { dirname, join } from "pathe";
import fs from "node:fs";
import { createSharedState } from "@vitejs/devtools-kit/utils/shared-state";
import { homedir } from "node:os";
//#region src/node/storage.ts
function createStorage(options) {
	const { mergeInitialValue = (initialValue, savedValue) => ({
		...initialValue,
		...savedValue
	}), debounce: debounceTime = 100 } = options;
	let initialValue = options.initialValue;
	if (fs.existsSync(options.filepath)) try {
		const savedValue = JSON.parse(fs.readFileSync(options.filepath, "utf-8"));
		initialValue = mergeInitialValue ? mergeInitialValue(options.initialValue, savedValue) : savedValue;
	} catch (error) {
		console.warn(`[Vite DevTools] Failed to parse storage file: ${options.filepath}, falling back to defaults.`);
		console.warn(error);
		initialValue = options.initialValue;
	}
	const state = createSharedState({
		initialValue,
		enablePatches: false
	});
	state.on("updated", debounce((newState) => {
		fs.mkdirSync(dirname(options.filepath), { recursive: true });
		fs.writeFileSync(options.filepath, `${JSON.stringify(newState, null, 2)}\n`);
	}, debounceTime));
	return state;
}
//#endregion
//#region src/node/auth-revoke.ts
/**
* Revoke an auth token: remove from storage and notify all connected clients
* using this token that they are no longer trusted.
*/
async function revokeAuthToken(context, storage, token) {
	storage.mutate((state) => {
		delete state.trusted[token];
	});
	const rpcHost = context.rpc;
	if (!rpcHost._rpcGroup) return;
	const affectedSessionIds = /* @__PURE__ */ new Set();
	for (const client of rpcHost._rpcGroup.clients) if (client.$meta.clientAuthToken === token) {
		affectedSessionIds.add(client.$meta.id);
		client.$meta.isTrusted = false;
		client.$meta.clientAuthToken = void 0;
	}
	if (affectedSessionIds.size === 0) return;
	await rpcHost.broadcast({
		method: "devtoolskit:internal:auth:revoked",
		args: [],
		filter: (client) => affectedSessionIds.has(client.$meta.id)
	});
}
//#endregion
//#region src/node/context-internal.ts
const internalContextMap = /* @__PURE__ */ new WeakMap();
function getInternalContext(context) {
	if (!internalContextMap.has(context)) {
		const storage = createStorage({
			filepath: join(homedir(), ".vite/devtools/auth.json"),
			initialValue: { trusted: {} }
		});
		const internalContext = {
			storage: { auth: storage },
			revokeAuthToken: (token) => revokeAuthToken(context, storage, token)
		};
		internalContextMap.set(context, internalContext);
	}
	return internalContextMap.get(context);
}
//#endregion
export { createStorage as n, getInternalContext as t };
